File: web/muse/tests/test_campaign_parser.py
import os
from django.test import TestCase
from pathlib import Path

from yaml.scanner import ScannerError

from muse.campaign_parser import CampaignParser
from muse.models import Server, Topic


class CampaignParserTest(TestCase):

    def setUp(self) -> None:
        super().setUp()
        self.parser = CampaignParser()

    def test_no_errors_in_good_file(self):
        yaml_file = os.path.dirname(os.path.abspath(__file__)) + '/data/valid_campaign.yaml'
        yaml = Path(yaml_file).read_text()
        is_ok = self.parser.verify(yaml)
        self.assertTrue(is_ok)

    def test_invalid_yaml(self):
        yaml = """entry:
            title: Entry
            text: should 
            not 
            be 
            here
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertIsInstance(self.parser.last_error, ScannerError)

    def test_no_extra_keys(self):
        yaml = """entry:
            title: Entry
            text: some text
            bad_value: should not be here
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry has invalid property bad_value')

    def test_alias_must_be_an_array(self):
        yaml = """entry:
            title: Entry
            text: some text
            aliases: some
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, f'aliases must be a list for entry')

    def test_text_max_length_2000(self):
        text = 'a1' * 1001
        yaml = f"""entry:
            title: Entry
            text: {text}
            aliases: 
                - some
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'text for entry exceeds max length of 2000')

    def test_key_max_length_100(self):
        key = 'a1' * 51
        yaml = f"""{key}:
            title: Entry
            text: This is the text
            aliases: 
                - some
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, f'{key} max length of 100 exceeded')

    def test_slug_max_length_100(self):
        slug = 'a1' * 51
        yaml = f"""entry:
            title: Entry
            wiki_slug: {slug}
            text: This is the text
            aliases: 
                - some
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, f'wiki_slug for entry exceeds max length of 100')

    def test_alias_max_length_100(self):
        alias = 'a1' * 51
        yaml = f"""entry:
            title: Entry
            text: This is the text
            aliases: 
                - {alias}
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, f'alias {alias} for entry exceeds max length of 100')

    def test_parent_max_length_100(self):
        parent = 'a1' * 51
        yaml = f"""entry:
            title: Entry
            text: This is the text
            parent: {parent}
            aliases: 
                - abc
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'parent for entry exceeds max length of 100')

    def test_key_cannot_start_with_hyphen(self):
        yaml = """'-entry':
            title: Entry
            text: some text
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, '-entry cannot start with a -')

    def test_text_is_required(self):
        yaml = """'entry':
            title: Entry
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry has no text')

        yaml = """'entry':
            title: Entry
            text:
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry / text invalid data type')

    def test_title_is_required(self):
        yaml = """'entry':
            text: Entry
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry has no title')

        yaml = """'entry':
            title:
            text: entry
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry has a blank title')

    def test_aliases_cannot_start_with_hyphen(self):
        yaml = """entry:
            title: Entry
            text: some text
            aliases:
                - '-one'
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'alias -one for entry cannot start with a -')

    def test_title_must_be_string(self):
        yaml = """entry:
            title:
                - the title
            text: some text
            page: 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry / title invalid data type')

    def test_page_must_be_string(self):
        yaml = """entry:
            title: the title
            text: some text
            page:
                - 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry / page invalid data type')

    def test_key_cannot_start_with_space(self):
        yaml = """" entry":
            title: the title
            text: some text
            page:
                - 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, ' entry cannot start with a space')

    def test_key_cannot_end_with_space(self):
        yaml = """"entry ":
            title: the title
            text: some text
            page:
                - 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry  cannot end with a space')

    def test_key_cannot_contain_punctuation(self):
        yaml = """"ent.ry":
            title: the title
            text: some text
            page:
                - 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'ent.ry cannot contain punctuation')

    def test_key_cannot_contain_multiple_spaces(self):
        yaml = """"my  entry":
            title: the title
            text: some text
            page:
                - 17
            parent: the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'my  entry cannot have multiple spaces')

    def test_parent_must_be_string(self):
        yaml = """entry:
            title: the title
            text: some text
            page: 17
            parent:
                - the parent
            aliases:
                - two
                - three
        """
        is_ok = self.parser.verify(yaml)
        self.assertFalse(is_ok)
        self.assertEqual(self.parser.last_error, 'entry / parent invalid data type')

    def test_save_entries(self):
        server = Server.objects.create(discord_id='test134', name='testserver')
        yaml_file = os.path.dirname(os.path.abspath(__file__)) + '/data/valid_campaign.yaml'
        yaml = Path(yaml_file).read_text()
        is_ok = self.parser.save(server.id, yaml)
        self.assertTrue(is_ok)
        self.assertEqual(Topic.objects.filter(server=server).count(), 6)
        self.assertEqual(Topic.objects.filter(server=server, alias_for=None).count(), 2)

File: bot/tests/testCampaignParser.js
"use strict"

const path = require('path');
require('dotenv').config({ path: path.resolve(__dirname, '../.env') });

const chai = require('chai');
chai.should();
const BasePersonality = require("../personalities/base");
const fs = require("fs");
const campaignParser = require("../campaignParser");

describe('Test campaign file verifier', function () {
  it('valid file parses ok', async() => {
    const yamlFile = path.resolve(__dirname, 'data', 'valid_campaign.yaml');
    let file = fs.readFileSync(yamlFile, 'utf8');
    const result = await campaignParser(file);
    chai.expect(result.error).to.be.null;
  });

  it('key cannot start with a hyphen', async() => {
    const entry = `'-Entry':
    text: some text
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('-Entry cannot start with a -');
  });

  it('key cannot start with a space', async() => {
    const entry = `' Entry':
    text: some text
    `;
    const result = await campaignParser(entry);
    result.error.should.equal(' Entry cannot start with a space');
  });

  it('key cannot end with a space', async() => {
    const entry = `'Entry ':
    text: some text
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('Entry  cannot end with a space');
  });

  it('key cannot contain multiple spaces', async() => {
    const entry = `'Entry  sample':
    text: some text
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('Entry  sample cannot have multiple spaces');
  });

  it('key cannot exceed max length', async() => {
    const key = 'a'.repeat(101);
    const entry = `'${key}':
    text: some text
    `;
    const result = await campaignParser(entry);
    result.error.should.contain('exceeds max length of');
  });

  it('entry must have text', async() => {
    const entry = `Sample:
    page: 19
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('Sample has no text');
  });

  it('entries must have valid keys', async() => {
    const entry = `Sample:
    text: some text
    bad_value: should not be here
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('Sample has an invalid attribute: bad_value');
  });

  it('aliases must be a list', async() => {
    let entry = `Sample:
    aliases: other
    text: some text
    `;
    let result = await campaignParser(entry);
    result.error.should.equal('Sample aliases must be a list');

    entry = `Sample:
    aliases:
        - other
    text: some text
    `;
    result = await campaignParser(entry);
    chai.expect(result.error).to.be.null;
  });

  it('aliases cannot start with hyphen', async() => {
    const entry = `'Entry':
    text: some text
    aliases:
      - two
      - '-one'
      - three
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('-one cannot start with a -');
  });

  it('only integers and strings allowed', async() => {
    const entry = `sample:
    text: 132.5
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('sample text is an invalid data type');
  });

  it('text does not exceed max length', async() => {
    const t = 'a'.repeat(4001);
    const entry = `Long:
    text: ${t}
    aliases:
      - two
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('Long text exceeds max length of 2000');
  });

  it('wiki_slug does not exceed max length', async() => {
    const t = 'a'.repeat(4001);
    const entry = `Long:
    text: test text
    wiki_slug: ${t}
    aliases:
      - two
    `;
    const result = await campaignParser(entry);
    result.error.should.equal('long text exceeds max length of 2000');
  });

});
